lib/rules/no-unused-services.js

```diff
@@ -0,0 +1,135 @@
+          unusedServicesMapping[name] = node;
+        }
+      },
+      // @service('bar') foo;
+      ClassProperty(node) {
+        if (emberUtils.isInjectedServiceProp(node)) {
+          const name = node.key.name;
+          unusedServicesMapping[name] = node;
+        }
+      },
+      // this.foo...
+      MemberExpression(node) {
+        if (isEmpty(unusedServicesMapping) || !propertyGetterUtils.isSimpleThisExpression(node)) {
+          return;
+        }
+        const name = node.property.name;
+        delete unusedServicesMapping[name];
+      },
+      // 1. get(this, 'foo...');
+      // 2. getProperties(this, [..., 'foo..', ...]); or getProperties(this, ..., 'foo..', ...);
+      // 3. this.get('foo...');
+      // 4. this.getProperties([..., 'foo..', ...]); or this.getProperties(..., 'foo..', ...);
+      CallExpression(node) {
+        if (isEmpty(unusedServicesMapping)) {
+          return;
+        }
+
+        if (node.callee.name === 'get' && types.isThisExpression(node.arguments[0])) {
+          // 1. get(this, 'foo...');
+          const name = splitValue(node.arguments[1].value);
+          delete unusedServicesMapping[name];
+        } else if (
+          node.callee.name === 'getProperties' &&
+          types.isThisExpression(node.arguments[0])
+        ) {
+          // 2. getProperties(this, [..., 'foo..', ...]); or getProperties(this, ..., 'foo..', ...);
+          const argArray = types.isArrayExpression(node.arguments[1])
+            ? node.arguments[1].elements
+            : node.arguments.slice(1);
+          for (const elem of argArray) {
+            const name = splitValue(elem.value);
+            delete unusedServicesMapping[name];
+          }
+        } else if (propertyGetterUtils.isThisGetCall(node)) {
+          // 3. this.get('foo...');
+          const name = splitValue(node.arguments[0].value);
+          delete unusedServicesMapping[name];
+        } else if (
+          types.isThisExpression(node.callee.object) &&
+          node.callee.property.name === 'getProperties'
+        ) {
+          // 4. this.getProperties([..., 'foo..', ...]); or this.getProperties(..., 'foo..', ...);
+          const argArray = types.isArrayExpression(node.arguments[0])
+            ? node.arguments[0].elements
+            : node.arguments;
+          for (const elem of argArray) {
+            const name = splitValue(elem.value);
+            delete unusedServicesMapping[name];
+          }
+        }
+      },
+      // const { foo, ... } = this;
+      VariableDeclarator(node) {
+        if (isEmpty(unusedServicesMapping)) {
+          return;
+        }
+
+        if (types.isThisExpression(node.init) && types.isObjectPattern(node.id)) {
+          for (const property of node.id.properties) {
+            delete unusedServicesMapping[property.key.name];
+          }
+        }
+      },
+      // Collect all the instances and report as needed
+      'Program:exit': () => {
```

Comment:
This rule currently assumes that a file may only contain a single class. It looks like this would currently be a valid test case, when actually a violation should be reported on the `myService` injection.

```js
class Class1 {
  @service() myService; // this is unused
}

class Class2 {
  someFunction() {
    this.myService.foo();
  }
}
```

Similarly, the rule doesn't handle nested classes:

```js
class Class1 {
  @service() myService; // this is unused

  someFunction1() {
    class Class2 {
      someFunction2() {
        this.myService.foo(); 
      }
    } 
  }
}
```

I think the ideal way to solve this would be to use the `classStack` approach from the `no-assignment-of-untracked-properties-used-in-tracking-contexts`. We maintain a stack to keep track of the classes we're inside. Each time we exit a class, we can check what service injections were unused in that class, and report violations.

Reply:
Oh nice! Thanks for pointing that out :) I'll take a look at `no-assignment-of-untracked-properties-used-in-tracking-contexts` and see how I can solve this

Reply:
Added the implementation + some tests! Let me know what you think 😄